server/src/QRServer/db/connector.py

```diff
@@ -387,6 +397,77 @@ async def _generate_ranking(self, start_date: datetime, end_date: datetime) -> L
             ))
         return ranking_entries
 
+    async def get_user_rating(self, user_id: str, month: int, year: int) -> Optional[UserRating]:
+        c = await self.conn.cursor()
+        await c.execute(
+            "select"
+            " rating,"
+            " revision"
+            " from user_ratings"
+            " where user_id = ?"
+            " and month = ?"
+            " and year = ?",
+            (user_id, month, year)
+            )
+        row = await c.fetchone()
+        if row:
+            return UserRating(user_id, month, year, rating=row[0], revision=row[1])
+        return None
+
+    async def update_users_rating(self, winner_id: str, loser_id: str, month: int, year: int):
+        # CONSTS
+        RETRIES = 3
+        for _ in range(RETRIES):
+            winner = await self.get_user_rating(winner_id, month, year)
+            winner_exists = bool(winner)
+            if not winner_exists:
+                winner = UserRating(winner_id, month, year)
+
+            loser = await self.get_user_rating(loser_id, month, year)
+            loser_exists = bool(loser)
+            if not loser_exists:
+                loser = UserRating(loser_id, month, year)
+
+            new_winner_rating, new_loser_rating = utils.calculate_new_ratings(winner.rating, loser.rating)
+
+            c = await self.conn.cursor()
+
+            if not winner_exists:
+                await c.execute(
+                    "insert into user_ratings (user_id, month, year, revision, rating) values (?, ?, ?, ?, ?)", (
+                        winner.user_id, month, year, winner.revision, new_winner_rating
+                    )
+                )
+            else:
+                await c.execute(
+                    "update user_ratings set rating = ?, revision = ?"
+                    " where user_id = ? and month = ? and year = ? and revision = ? returning *", (
+                        new_winner_rating, winner.revision + 1, winner_id, month, year, winner.revision
+                    )
+                )
+
+                row = await c.fetchone()
+                if not row:
+                    self.conn.rollback()
```

Comment:
Is this rollback and retry behavior tested? I think it should be

Reply:
it is now